sdv/single_table/utils.py

```diff
@@ -32,10 +32,19 @@ def detect_discrete_columns(metadata, data):
     """
     discrete_columns = []
     for column in data.columns:
-        # Numerical and datetime columns never get preprocessed into categorical ones
-        if column in metadata.columns and \
-                metadata.columns[column]['sdtype'] in ['numerical', 'datetime']:
-            continue
+        if column in metadata.columns:
+            sdtype = metadata.columns[column]['sdtype']
+            # Numerical and datetime columns never get preprocessed into categorical ones
+            if sdtype in ['numerical', 'datetime']:
+                continue
+
+            elif sdtype in ['categorical', 'boolean']:
+                transformer = transformers.get(column)
+                if transformer and transformer.get_output_sdtypes().get(column) == 'float':
+                    continue
+
+                discrete_columns.append(column)
+                continue
```

Comment:
logic looks good but I think we don't need the logic below anymore. Before it was trying to look at columns marked as categorical in the metadata, and decide if we should treat them as discrete based on the unique values. But now since we actually are using the transformers to see what is categorical or not, I don't think it's still useful

Reply:
The underlying logic should remain unchanged for the following reason: When the `column` in question is missing from the metadata but is generated by a `transformer`, we rely on the `data` to determine whether it represents a discrete or numerical column. 

For instance, consider the case where we produce the `is_nan` column using the `model_missing_values`. Despite the column being numerical, we categorize it as discrete, given that it only contains two numerical categories.